File: fix_lints.py
import os
import re

def fix_markdown(filepath):
    with open(filepath, 'r', encoding='utf-8') as f:
        lines = f.readlines()

    out_lines = []
    
    # Pre-computation
    for i, line in enumerate(lines):
        # Fix MD009 (Trailing spaces)
        line = line.rstrip() + '\n'
        
        # FIx MD030 (List marker space - multiple spaces after asterisk or dash)
        # matches "-   " or "*   " at the start of line with spaces before it
        line = re.sub(r'^(\s*[-*]) {2,}', r'\1 ', line)

        out_lines.append(line)
        
    final_lines = []
    for i, line in enumerate(out_lines):
        # Fix MD022 (Headings surrounded by blank lines)
        if re.match(r'^\#{1,6}\s', line):
            # Ensure blank line before
            if i > 0 and out_lines[i-1].strip() != '' and not re.match(r'^\s*$', final_lines[-1]):
                final_lines.append('\n')
            final_lines.append(line)
            # Ensure blank line after wait we inject next loop if needed, but actually we can just insert it here
            if (i+1 < len(out_lines)) and out_lines[i+1].strip() != '':
                 final_lines.append('\n')
            continue
            
        # Fix MD032 (Lists surrounded by blank lines)
        is_list = bool(re.match(r'^\s*[-*]\s', line))
        was_list = i > 0 and bool(re.match(r'^\s*[-*]\s', out_lines[i-1]))
        if is_list and not was_list:
            if final_lines and final_lines[-1].strip() != '':
                final_lines.append('\n')

        # Fix MD060 (Table pipe style space parsing)
        if "|" in line and "-|-" in line.replace(" ", ""):
            line = line.replace("|-", "| -").replace("-|", "- |")
            line = re.sub(r'\|\s*\-\s*\|', '| --- |', line)
            
        final_lines.append(line)

    with open(filepath, 'w', encoding='utf-8') as f:
        f.writelines(final_lines)

    print(f"Fixed {filepath}")

for root, dirs, files in os.walk('.'):
    for fn in files:
        if fn.endswith('.md'):
            fix_markdown(os.path.join(root, fn))

File: fix_lints.js
const fs = require('fs');
const path = require('path');

function walk(dir) {
    let results = [];
    const list = fs.readdirSync(dir);
    list.forEach(function(file) {
        file = dir + '/' + file;
        const stat = fs.statSync(file);
        if (stat && stat.isDirectory()) {
            if (!file.includes('.git')) {
                results = results.concat(walk(file));
            }
        } else {
            if (file.endsWith('.md')) {
                results.push(file);
            }
        }
    });
    return results;
}

const files = walk('.');

files.forEach(file => {
    let lines = fs.readFileSync(file, 'utf8').split('\n');
    let outLines = [];
    
    // Fix Trailing Spaces and List Marker spacing
    lines.forEach((line) => {
        line = line.replace(/[ \t]+$/, ''); // trailing space
        line = line.replace(/^(\s*[-*])\s{2,}/, '$1 '); // list marker space
        outLines.push(line);
    });
    
    let finalLines = [];
    outLines.forEach((line, i) => {
        // Fix Headings format
        if (line.match(/^#{1,6}\s/)) {
            if (i > 0 && outLines[i-1].trim() !== '') {
                finalLines.push('');
            }
            finalLines.push(line);
            if (i + 1 < outLines.length && outLines[i+1].trim() !== '') {
                finalLines.push('');
            }
            return;
        }
        
        // Blank lines before lists
        let isList = !!line.match(/^\s*[-*]\s/);
        let wasList = i > 0 && !!outLines[i-1].match(/^\s*[-*]\s/);
        if (isList && !wasList && finalLines.length > 0 && finalLines[finalLines.length-1].trim() !== '') {
            finalLines.push('');
        }
        
        // Fix Table style MD060
        if (line.includes('|') && line.replace(/\s/g, '').includes('-|-')) {
            line = line.replace(/\|-/g, '| -').replace(/-\|/g, '- |').replace(/\|\s*-\s*\|/g, '| --- |');
            // Ensure no tightly packed pipe segments like |---|---|
            let parts = line.split('|');
            parts = parts.map(p => {
               if (p.trim().match(/^-+$/)) {
                   return ' --- ';
               }
               return p;
            });
            line = parts.join('|');
            // Remove multiple leading/trailing borders if duplicated
            if (line.startsWith('|  --- ')) line = '| --- ';
        }
        
        finalLines.push(line);
    });
    
    fs.writeFileSync(file, finalLines.join('\n'));
});
